internal/validators/oci_validator.go

```diff
@@ -0,0 +1,318 @@
+	if err != nil {
+		errMsg := "failed to setup authentication"
+		s.log.V(0).Info(errMsg, "error", err.Error(), "rule", rule.Name(), "host", rule.Host, "auth", rule.Auth)
+		s.updateResult(vr, []error{err}, errMsg, rule.Name())
+		return vr, nil
+	}
+
+	errs := make([]error, 0)
+	details := make([]string, 0)
+	ctx := context.Background()
+	if len(rule.Artifacts) == 0 {
+		errMsg := "failed to validate repositories in registry"
+		d, err := validateRepos(ctx, rule.Host, opts, vr)
+		if err != nil {
+			errs = append(errs, err)
+			s.log.V(0).Info(errMsg, "error", err.Error(), "rule", rule.Name(), "host", rule.Host)
+		}
+		if len(d) > 0 {
+			details = append(details, d...)
+		}
+
+		s.updateResult(vr, errs, errMsg, rule.Name(), details...)
+		return vr, nil
+	}
+
+	errMsg := "failed to validate artifact in registry"
+	for _, artifact := range rule.Artifacts {
+		ref, err := generateRef(rule.Host, artifact.Ref, vr)
+		if err != nil {
+			details = append(details, fmt.Sprintf("failed to generate reference for artifact %v/%v", rule.Host, artifact.Ref))
+			errs = append(errs, err)
+			s.log.V(0).Info(errMsg, "error", err.Error(), "rule", rule.Name(), "host", rule.Host, "artifact", artifact)
+			continue
+		}
+
+		detail, err := validateReference(ref, artifact.Download, opts)
+		if err != nil {
+			errs = append(errs, err)
+			s.log.V(0).Info(errMsg, "error", err.Error(), "rule", rule.Name(), "host", rule.Host, "artifact", artifact)
+		}
+		if detail != "" {
+			details = append(details, detail)
+		}
+	}
+	s.updateResult(vr, errs, errMsg, rule.Name(), details...)
+	return vr, nil
+}
+
+func generateRef(registry, artifact string, vr *types.ValidationResult) (name.Reference, error) {
+	if strings.Contains(artifact, "@sha256:") {
+		return name.NewDigest(fmt.Sprintf("%s/%s", registry, artifact))
+	}
+
+	if !strings.Contains(artifact, ":") {
+		vr.Condition.Details = append(vr.Condition.Details, fmt.Sprintf("artifact %v does not contain a tag or digest, defaulting to \"latest\" tag", artifact))
+	}
+
+	return name.NewTag(fmt.Sprintf("%s/%s", registry, artifact))
+}
+
+// validateArtifact validates a single artifact within a registry. This function is to be used when a path to a repo or an individual artifact is provided
+func validateReference(ref name.Reference, download bool, opts []remote.Option) (string, error) {
+	// validate artifact existence by issuing a HEAD request
+	_, err := remote.Head(ref, opts...)
+	if err != nil {
+		return fmt.Sprintf("failed to get descriptor for artifact %v", ref.Name()), err
+	}
+
+	// TODO: add signature verification here
+
+	if download {
+		// download image without storing it on disk
+		img, err := remote.Image(ref, opts...)
+		if err != nil {
+			return fmt.Sprintf("failed to download artifact %v", ref.Name()), err
+		}
+
+		err = validate.Image(img)
+		if err != nil {
+			return fmt.Sprintf("failed validation for artifact %v", ref.Name()), err
+		}
```

Comment:
This was a nice find in the `go-containerregistry` library. Lets them do the heavy lifting when it comes to image validation